lib/pure/asyncfutures.nim

```diff
@@ -23,6 +23,8 @@ type
     callbacks: CallbackList
 
     finished: bool
+    cancelled: bool
+    recursiveFuture*: FutureBase        ## Stored future for recursive cancellation
```

Comment:
I finally got to this and it's midnight so I'm likely missing something. But my initial impressions from reading this are that this looks awfully like emulating exception propagation logic, couldn't we simply create a `FutureCancelled` exception and have it propagate up the futures?

The reason I ask this is because if we can then it seems like it would be an even simpler and cleaner addition to what we already have.

Reply:
We need to store recursive futures because the iterator loop is not storing them anywhere, just adding callbacks to them and that's all. The idea from chronos to store all the futures in hashtable seems too complicated and not so needed to me. The cancelled flag is needed only for those recursive futures to stop immediately without any error.

The idea of replacing the cancelled flag to failure with FutureCancelled might lead to situation when the recurrent future will raise the cancelled exception while it is not realy cancelled. The difference with chronos is that my implementation is not cancelling the future - it is completing or failing it, so we just break the iterator loop and call all the callbacks which are connected to that future with setting the value or error for that future, but to completely stop we need to stop all the recurrent futures silently (IMO) without raising the cancelled error because this could lead to an exception mess and stopping of the application.

Reply:
> The idea of replacing the cancelled flag to failure with FutureCancelled might lead to situation when the recurrent future will raise the cancelled exception while it is not realy cancelled.

How could it? We pretty much control when the Future exception gets re-raised.

Reply:
> How could it? We pretty much control when the Future exception gets re-raised.

The cancellation of all subsequent futures will not be silent, they all will get this FutureCancelled exception, so they all will reraise it. Not sure that it is a good behaviour, especially when this exception will return to a parent one which is not cancelled at all, but just completed or failed. BTW, in this case the failure of a parent future might be replaced with a failure of a subsequent exception, which is not good too.

If you insist on replacing this flag to a FutureCancelled exception then we should separate the parent future and all subsequent ones by some flag also to not allow such a behaviour which I mentioned before.